contributors/design-proposals/lifecycle-hooks.md

```diff
@@ -0,0 +1,152 @@
+# Lifecycle Hooks
+
+**Author**: @tnozicka
+
+**Status**: Proposal
+
+**RFC Issue**: https://github.com/kubernetes/kubernetes/issues/14512
+
+## Abstract
+The intent of this proposal is to support lifecycle hooks in all objects having rolling/recreate update; currently those are Deployments, StatefulSets and DaemonSets. 
+Lifecycle hooks should allow users to better customize updates (rolling/recreate) without having to write their own controllers. 
+
+## Lifecycle Hooks
+Lifecycle hook is a Job that will get triggered by the update reaching certain progress point (like 50%). 
+The progress point is determined by the ratio of new.availableReplicas to the declared replicas or by explicitly stating the number of new.availableReplicas. 
+The API is open for other types of triggering conditions in the future. 
+
+### Previous Implementations
+Lifecycle hooks are already implemented in OpenShift but we want to enhance them and implement them natively in Kubernetes.
+ - [OpenShift proposal](https://github.com/openshift/origin/blob/master/docs/proposals/post-deployment-hooks.md)
+ - [OpenShift docs](https://docs.openshift.org/latest/dev_guide/deployments/deployment_strategies.html#lifecycle-hooks)
+
+### Previous Kubernetes Proposals
+ - https://github.com/kubernetes/kubernetes/pull/33545
+
+### Use Cases
+The most common use case for lifecycle hooks is to have pre, mid and/or post hook. 
+It is mostly used to run some kind of acceptance check in the middle and/or at the end to fully verify the update is working as expected and rollback if it isn't. 
+The acceptance check may be time consuming and more thorough than what readiness and liveness probes are intended for. 
+You can also notify external services from them, migrate database in the middle of an update, send messages to IRC channel or do anything else.
+
+[A short demo](https://youtu.be/GVNTm_K43iI) simulating lifecycle hooks using auto-pausing that has been presented at SIG-Apps meeting on August 21, 2017.
+
+
+## API Objects
+### New Objects
+```go
+type HookTriggerType string
+
+const (
+	ProgressAvailableHookTriggerType HookTriggerType = "ProgressAvailable"
+)
+
+type ProgressAvailableParams struct {
+    // ProgressPoint specifies the point during update when the hook should be triggered.
+    // Accepts both the number of new pods available or a percentage value representing the ratio
```

Comment:
If we are going to implement something like this then isn't autopausing an easier thing to implement and handle arbitrary workflow execution (start a job or really anything else) outside of the core controllers?

Reply:
> If we are going to implement something like this then isn't autopausing an easier thing to implement

From the scaling point of view, the amount of work will be slightly less as this will be like setting a partition, only internally, without exposing it in the API (yet). We can as well expose the partition in API (auto-pausing) but I didn't want to mix it into this proposal.

> handle arbitrary workflow execution (start a job or really anything else) outside of the core controllers?

I'd prefer to see it in apps. I think this would be beneficial and helpful for the users - I don't want to decouple this and have it only in some distribution just because we can.